securedrop/source_app/main.py

```diff
@@ -37,26 +37,29 @@ def generate():
             return redirect(url_for('.lookup'))
 
         codename = generate_unique_codename()
-        session['codename'] = codename
         session['new_user'] = True
         return render_template('generate.html', codename=codename)
 
     @view.route('/create', methods=['POST'])
     def create():
-        filesystem_id = crypto_util.hash_codename(session['codename'])
+        codename = request.form['codename']
+        filesystem_id = crypto_util.hash_codename(codename)
 
         source = Source(filesystem_id, crypto_util.display_id())
         db_session.add(source)
         try:
             db_session.commit()
-        except IntegrityError as e:
+        except IntegrityError:
             db_session.rollback()
-            current_app.logger.error(
-                "Attempt to create a source with duplicate codename: %s" %
-                (e,))
+            # Issue #2386: abort if the codename already exists
+            flash(gettext('There was an error creating your user. A new '
+                          'codename was generated. Please try again.'),
+                  'error')
```

Comment:
The desired logic here should be to log the event and just abort with a 500 if the duplicate codename event occurs since this is defense in depth and not a path users should encounter? (Aborting with 500 is what is described in #2386).